model/pipeline/src/main/proto/beam_runner_api.proto

```diff
@@ -963,16 +963,39 @@ message SideInput {
   SdkFunctionSpec window_mapping_fn = 3;
 }
 
-// An environment for executing UDFs. Generally an SDK container URL, but
-// there can be many for a single SDK, for example to provide dependency
-// isolation.
+// An environment for executing UDFs. By default, an SDK container URL, but
+// can also be a process forked by a command, or an externally managed process.
 message Environment {
 
-  // (Required) The URL of a container
-  //
-  // TODO: reconcile with Fn API's DockerContainer structure by
-  // adding adequate metadata to know how to interpret the container
-  string url = 1;
+  // (Required) The URN of the payload
+  string urn = 1;
+
+  // (Optional) The data specifying any parameters to the URN. If
+  // the URN does not require any arguments, this may be omitted.
+  bytes payload = 2;
+}
+
+message StandardEnvironments {
+  enum Environments {
+    DOCKER = 0 [(beam_urn) = "beam:env:docker:v1"];
+
+    PROCESS = 1 [(beam_urn) = "beam:env:process:v1"];
+
+    EXTERNAL = 2 [(beam_urn) = "beam:env:external:v1"];
+  }
```

Comment:
EMBEDDED was another option mentioned on the discussion thread. I think it may also be helpful to add a line explaining what each option means.

Reply:
@angoenka this remains to be addressed

Reply:
Do you think we should add Embedded yet? It seems the least fleshed out and will only support Java pipelines.

Reply:
Correct, so perhaps we should also get rid of `EXTERNAL` then?